precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+		let refund_address = input.read::<Address>()?.0;
+		let fee: BalanceOf<Runtime> = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		let num_words = input.read::<u8>()?;
+		let blocks_after_current = input.read::<u64>()?;
+		handle.record_cost(
+			REQUEST_RANDOMNESS_ESTIMATED_COST + RuntimeHelper::<Runtime>::db_read_gas_cost(),
+		)?;
+		let current_block_number: u64 = <frame_system::Pallet<Runtime>>::block_number()
+			.try_into()
+			.map_err(|_| revert("block number overflowed u64"))?;
+		let requested_block_number = blocks_after_current
+			.checked_add(current_block_number)
+			.ok_or(error("addition result overflowed u64"))?
+			.try_into()
+			.map_err(|_| revert("u64 addition result overflowed block number type"))?;
+		let request = Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			num_words,
+			salt,
+			info: RequestType::Local(requested_block_number),
+		};
+
+		let request_id: U256 = Pallet::<Runtime>::request_randomness(request)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?
+			.into();
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: EvmDataWriter::new().write(request_id).build(),
+		})
+	}
+	/// Fulfill a randomness request due to be fulfilled
+	fn fulfill_request(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let request_id: u64 = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| error("Input RequestId overflows u64 type set in Pallet"))?;
+		// read all the inputs
+		let pallet_randomness::FulfillArgs {
+			request,
+			deposit,
+			randomness,
+		} = Pallet::<Runtime>::prepare_fulfillment(request_id)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?;
+		// check that randomness can be provided
+		ensure_can_provide_randomness::<Runtime>(
+			handle.code_address(),
+			request.gas_limit,
+			request.fee,
+			FULFILLMENT_OVERHEAD_ESTIMATED_COST,
+		)?;
+		// get gas before subcall
+		let before_remaining_gas = handle.remaining_gas();
+		provide_randomness(
+			handle,
+			request_id,
+			request.gas_limit,
+			request.contract_address.clone().into(),
+			randomness.into_iter().map(|x| H256(x)).collect(),
+		)?;
+		// get gas after subcall
+		let after_remaining_gas = handle.remaining_gas();
+		let gas_used: U256 = before_remaining_gas
+			.checked_sub(after_remaining_gas)
+			.ok_or(revert("Before remaining gas < After remaining gas"))?
+			.into();
+		// cost of execution is before_remaining_gas less after_remaining_gas
+		let cost_of_execution: BalanceOf<Runtime> = gas_used
+			.checked_mul(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
+			.ok_or(revert("Multiply gas used by base fee overflowed"))?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
```

Comment:
What happens to the subcall state changes when we `revert()` here (they should revert, too, right?)

Is the assumption with this error handling that it's not a case we can possibly hit? (I think the only way to do it is to have a misconfigured runtime with an impossibly high `base_fee_per_gas`).

Reply:
> What happens to the subcall state changes when we revert() here (it should revert, too, right?)

I don't think it will because it's a separate call. @nanocryk 

> Is the assumption with this error handling that it's not a case we can possibly hit? (I think the only way to do it is to have a misconfigured runtime with an impossibly high `base_fee_per_gas`).

Pretty much. I don't think it's better to panic here and we need the inner value. There is no infallible way to make these type conversions and we can't make them before the subcall.

Reply:
You could saturate instead. Not sure that's any better.

In this case, it's pretty important that a subcall reverts the whole thing, otherwise it could be possible to invoke the contract's callback more than once per request. (note that this should be impossible with our runtime configurations).